proposals/init.md

```diff
@@ -0,0 +1,582 @@
+adjusted to account for `init` members. This will need to be included in the 
+rule changes for non-mutating acess to `readonly` data.
+
+The IL verification rules will need to be broken into two parts: 
+
+1. Allowing `init` members to set a `readonly` field.
+1. Determining when an `init` member can be legally called.
+
+The first is a simple adjustmetn to the existing rules. The IL verifier can 
+be taught to recognize `init` members and from there it just needs to consider
+a `readonly` field to be settable on `this` in such a member.
+
+The second rule is more complicated. In the simple case of object initializers
+the rule is straight forward. It should be legal to call `init` members when 
+the result of a `new` expression is still on the stack. That is until the 
+value has been stored in a local, array element or field or passed as an
+argument to another method it will still be legal to call `init` members. This
+ensures that once the result of the `new` expression is published to a named
+identifier (other than `this`) then it will no longer be legal to call `init`
+members. 
+
+The more complicated case though is when we mix `init` members, object
+initializers and `await`. That can cause the newly craeted object to be
+temporarily hoisted into a state machine and hence put into a field.
+
+```cs
+var student = new Student() 
+{
+    Name = await SomeMethod()
+};
+```
+
+Here the result of `new Student()` will be hoised into a state machine as a 
+field before the set of `Name` occurs. The compiler will need to mark such
+hoisted fields in a way that the IL verifier understands they're not user 
+accessible and hence doesn't violate the intended semantics of `init`.
+
+## init members
+The `init` modifier could be extended to apply to all instance members. This 
+would generalize the concept of `init` during object construction and allow
+types to declare helper methods that could partipate in the construction 
+process to initialize `init` fields and properties.
+
+Such members would have all the restricions that an `init` accessor does
+in this design. The need is questionable though and this can be safely added
+in a future version of the language in a compatible manner.
+
+### Generate three accessors
+One potential implementation of `init` properties is to make `init` completely
+separate from `set`. That means that a property can potentially have three 
+different accessors: `get`, `set` and `init`.
+
+This has the potential advantage of allowing the use of modreq to enforce 
+correctness while maintaining binary compatibility. The implementation would
+roughly be the following:
+
+1. An `init` accessor is always emitted if there is a `set`. When not defined 
+by the developer it is simply a reference to `set`. 
+1. The set of a property in an object initializer will always use `init` if 
+present but fall back to `set` if it's missing.
+
+This means that a developer can always safely delete `init` from a property. 
```

Comment:
This assumes that the property also has a setter, right? For init-only properties this would not work, right?

Reply:
The rules were meant to mention that defining `init` or `set` would end requiring both were emitted. Will update.